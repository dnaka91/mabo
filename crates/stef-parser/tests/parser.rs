use std::{
    fmt::{self, Display},
    fs,
};

use insta::{assert_snapshot, glob, with_settings};
use miette::{Diagnostic, MietteHandler, MietteHandlerOpts, ReportHandler};
use stef_parser::Schema;

#[test]
fn parse_schema() {
    glob!("inputs/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str()).unwrap();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("parse", format!("{value:#?}"));
            assert_snapshot!("print", value.to_string());
        });
    });
}

#[test]
fn parse_invalid_schema() {
    struct Wrapper<'a>(&'a MietteHandler, &'a dyn Diagnostic);

    impl<'a> Display for Wrapper<'a> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            self.0.debug(self.1, f)
        }
    }

    let handler = MietteHandlerOpts::new()
        .color(false)
        .terminal_links(false)
        .width(120)
        .force_graphical(true)
        .build();

    glob!("inputs/invalid/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str()).unwrap_err();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error", Wrapper(&handler, &*value).to_string());
        });
    });
}
