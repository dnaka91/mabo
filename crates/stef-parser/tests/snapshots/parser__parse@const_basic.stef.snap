---
source: crates/stef-parser/tests/parser.rs
description: "const BOOL_TRUE: bool = true;\nconst BOOL_FALSE: bool = false;\nconst INT: u32 = 100;\nconst FLOAT: f64 = 5.5;\nconst STRING: string = \"value\";\nconst BYTES: bytes = [1, 2, 3];"
input_file: crates/stef-parser/tests/inputs/const_basic.stef
---
Schema {
    path: Some(
        "const_basic.stef",
    ),
    source: "const BOOL_TRUE: bool = true;\nconst BOOL_FALSE: bool = false;\nconst INT: u32 = 100;\nconst FLOAT: f64 = 5.5;\nconst STRING: string = \"value\";\nconst BYTES: bytes = [1, 2, 3];\n",
    comment: Comment(
        [],
    ),
    definitions: [
        Const(
            Const {
                comment: Comment(
                    [],
                ),
                name: Name {
                    value: "BOOL_TRUE",
                },
                ty: Type {
                    value: Bool,
                },
                value: Literal {
                    value: Bool(
                        true,
                    ),
                },
            },
        ),
        Const(
            Const {
                comment: Comment(
                    [],
                ),
                name: Name {
                    value: "BOOL_FALSE",
                },
                ty: Type {
                    value: Bool,
                },
                value: Literal {
                    value: Bool(
                        false,
                    ),
                },
            },
        ),
        Const(
            Const {
                comment: Comment(
                    [],
                ),
                name: Name {
                    value: "INT",
                },
                ty: Type {
                    value: U32,
                },
                value: Literal {
                    value: Int(
                        100,
                    ),
                },
            },
        ),
        Const(
            Const {
                comment: Comment(
                    [],
                ),
                name: Name {
                    value: "FLOAT",
                },
                ty: Type {
                    value: F64,
                },
                value: Literal {
                    value: Float(
                        5.5,
                    ),
                },
            },
        ),
        Const(
            Const {
                comment: Comment(
                    [],
                ),
                name: Name {
                    value: "STRING",
                },
                ty: Type {
                    value: String,
                },
                value: Literal {
                    value: String(
                        "value",
                    ),
                },
            },
        ),
        Const(
            Const {
                comment: Comment(
                    [],
                ),
                name: Name {
                    value: "BYTES",
                },
                ty: Type {
                    value: Bytes,
                },
                value: Literal {
                    value: Bytes(
                        [
                            1,
                            2,
                            3,
                        ],
                    ),
                },
            },
        ),
    ],
}
